### PR TITLE
Add support for tracked properties with proxied managed array

### PR DESCRIPTION
**Pull request by igorT**
Because ManagedArray is now behind a proxy, it does not pass the internal ember `isEmberArray` check that uses a WeakSet https://github.com/emberjs/ember.js/blob/21bd70c773dcc4bfe4883d7943e8a68d203b5bad/packages/%40ember/-internals/utils/lib/ember-array.ts, so we have to access '[]' when we access ManagedArrays

### CODE REVIEW

File: addon/base-record-array.js [hunk 1]
```diff
@@ -108,6 +112,7 @@ if (CUSTOM_MODEL_CLASS) {
     }
 
     objectAt(idx) {
+      get(this, '[]');
```
Comment (hjdivad) [inline addon/base-record-array.js:115]: this needs an explanatory comment

File: addon/base-record-array.js [hunk 2]
```diff
@@ -206,6 +211,7 @@ if (CUSTOM_MODEL_CLASS) {
     }
 
     get length() {
+      get(this, '[]');
```
Comment (hjdivad) [inline addon/base-record-array.js:214]: explanatory comment please

File: addon/model.js [hunk 3]
```diff
@@ -145,6 +146,16 @@ if (CUSTOM_MODEL_CLASS) {
...
 
+// If a value returned from unknownProperty is a ManagedArray we need to access '[]'
+// to keep parity with how ember treats arrays
+// see https://github.com/emberjs/ember.js/blob/3ce13cea235cde8a87d89473533c453523412764/packages/%40ember/-internals/metal/lib/property_get.ts#L136
+function accessBracketsIfArray(maybeArray) {
+  if (MANAGED_ARRAYS.has(maybeArray)) {
+    get(maybeArray, '[]');
+  }
+  return maybeArray;
+}
```
Comment (hjdivad) [inline addon/model.js:157]: Let's rename this to something like `readCachedValue(key)`

As it stands the naming is odd. You get usage like

```js
  if (key in this._cache) {
      return accessBracketsIfArray(this._cache[key]);
    }
```

So here the call site cares about the return value, but is calling a function named for its side-effects. This is a kind of abstraction leakage -- contrast to

```js
  if (key in this._cache) {
      return readCachedValue(this._cache, key);
    }
```

now the call site is asking for what it needs in order to return a value (the cached value) and the side effects necessary to entangle the array via ember's semantics is encapsulated in the implementation of `readCachedValue`.

File: tests/unit/model/native-access-arrays-test.js [hunk 4]
```diff
@@ -99,17 +99,6 @@ if (CUSTOM_MODEL_CLASS) {
...
-      let shiftedChapter = shiftedChapters.shift();
-
-      assert.equal(shiftedChapters[0].get('name'), 'The Boy Who Lived', `Array.shift works`);
-      assert.equal(shiftedChapters.length, 1, `Array.shift works`);
-      assert.equal(shiftedChapter.get('name'), 'The Vanishing Glass', 'Array.shift works');
-
-      shiftedChapters.unshift(shiftedChapter);
-      assert.equal(shiftedChapters[0].get('name'), 'The Vanishing Glass', `Array.unshift works`);
-      assert.equal(shiftedChapters.length, 2, `Array.unshift works`);
-
```
Comment (hjdivad) [inline tests/unit/model/native-access-arrays-test.js:112]: why are these removed?

File: tests/unit/model/native-access-arrays-test.js [hunk 5]
```diff
@@ -99,17 +99,6 @@ if (CUSTOM_MODEL_CLASS) {
       assert.equal(slicedChapters[0].get('name'), 'The Boy Who Lived', `Array.slice still works`);
```
Comment (hjdivad) [inline tests/unit/model/native-access-arrays-test.js:99]: why do we use `get` in these tests?

**Review by hjdivad — CHANGES_REQUESTED** [review]

File: tests/unit/model/native-access-arrays-test.js (hunk 5 above)
Comment (igorT) [reply tests/unit/model/native-access-arrays-test.js:99]: To not depend on `useNativeProperties` being true

File: tests/unit/model/native-access-arrays-test.js (hunk 4 above)
Comment (igorT) [reply tests/unit/model/native-access-arrays-test.js:112]: These have been moved to https://github.com/hjdivad/ember-m3/pull/1291/files/ea086c142eeb5656dcd4f8135c5038fa4ffaeae7#diff-ca4b95cfbe6d1d2a7c6d1b9d8d44b4593ee6b0470174d1e49b004bfd14bb65b1R171 to simplify the tests

File: tests/unit/model/native-access-arrays-test.js (hunk 5 above)
Comment (hjdivad) [reply tests/unit/model/native-access-arrays-test.js:99]: do we test the common case of using both of these?

File: addon/base-record-array.js (hunk 1 above)
Comment (igorT) [reply addon/base-record-array.js:117]: done

File: addon/base-record-array.js (hunk 2 above)
Comment (igorT) [reply addon/base-record-array.js:218]: done

File: addon/model.js (hunk 3 above)
Comment (igorT) [reply addon/model.js]: done

File: tests/unit/model/native-access-arrays-test.js (hunk 5 above)
Comment (igorT) [reply tests/unit/model/native-access-arrays-test.js:99]: The tests added here use them together.

File: addon/model.js (hunk 3 above)
Comment (hjdivad) [reply addon/model.js]: what's the reason for preferring `readCachedValue(value)` over the API I suggested above?

File: tests/unit/model/native-access-arrays-test.js (hunk 5 above)
Comment (hjdivad) [reply tests/unit/model/native-access-arrays-test.js:99]: it's just `'mutating reference arrays cause re-renders'` isn't it?  the other two use `get`.

I guess it's probably okay, I don't have any particular reason to think these two features don't work well together, I just find it a bit odd that we don't have many tests for the most common scenario

Comment (igorT) [reply tests/unit/model/native-access-arrays-test.js:99]: Happy to add more separately.

File: addon/model.js (hunk 3 above)
Comment (igorT) [reply addon/model.js]: Discussed on slack, updated the method now so we don't have a separate function call

**Review by hjdivad — APPROVED** [review]
LGTM please rebase with a clean  commit history before merging
